### ISSUE TITLE
Rule proposal: Enforce property reactivity for Vue 2

### ISSUE THREAD

**Issue by rezbyte** [vuejs/eslint-plugin-vue#1990]
**Please describe what the rule should do:**

Warn against adding or deleting properties of an object in data.
It should also warn against mutating an item via index, and modifying the length of an array.
Enforce the usage of `Vue.set`/`this.$set` & `Vue.delete`/`this.$delete`.

**What category should the rule belong to?**

[ ] Enforces code style (layout)
[X] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule should warn about:**

```js
export default {
  name: "component",

  data() {
    return {
      a: {},
    };
  },

  methods: {
    setB() {
      this.a.b = true;
    },
  },
};
```

```js
export default {
  name: "component",

  data() {
    return {
      a: {
        b: true,
      },
    };
  },

  methods: {
    deleteB() {
      delete this.a.b;
    },
  },
};

```

```js
export default {
  name: "component",

  data() {
    return {
      array: ["a","b","c"],
    };
  },

  methods: {
    setToD() {
      this.array[2] = "d";
    },
  },
};
```

```js
export default {
  name: "component",

  data() {
    return {
      array: ["a","b","c"],
    };
  },

  methods: {
    increaseLength() {
      this.array.length = 4;
    },
  },
};

```

**Additional context**

[Vue 2 can't make dynamic changes to an object's structure reactive as per the docs.](https://v2.vuejs.org/v2/guide/reactivity.html#Change-Detection-Caveats)

**Comment by alexanderturinske** [comment]
Since these APIs are [removed in Vue 3](https://v3-migration.vuejs.org/breaking-changes/#removed-apis), maybe we should enforce not using them now. 🤔 

`Global functions set and delete, and the instance methods $set and $delete. They are no longer required with proxy-based change detection.`

**Comment by FloEdelmann** [comment]
@rezbyte Adding new rules for Vue 2 doesn't really make sense at this time anymore.

@alexanderturinske Yes, adding a new rule `vue/no-deprecated-get-set` would make sense. Could you please open a new issue and link to this issue here, so we keep it clean?

**Comment by alexanderturinske** [comment]
Thank you for your guidance @FloEdelmann . I have created the new issue at https://github.com/vuejs/eslint-plugin-vue/issues/2526
